# <center> <div style="background-color:black; padding:30px 20px; text-align:center; border-bottom:4px solid red;">
  <h1 style="color:red; margin:0; font-size:50px; letter-spacing:2px;">
    <center> COVID-19 CoronaVirus
  </h1>
</div>

<img src="coronavirus.png" width="7000">

## <center> <h1 style="color:white; background-color:black; padding:8px;">Introduction</h1>
### COVID-19 is a respiratory disease caused by the SARS-CoV-2 coronavirus. It was first reported in Wuhan, China, in late December 2019, and the World Health Organization declared it a global pandemic in March 2020. This project uses a country-level COVID-19 dataset with 54,522 records covering 194 countries or regions, with one row for each country on each date, from 31 December 2019 to 6 November 2020. Each record contains the cumulative confirmed cases, deaths and recoveries, along with the new cases, deaths and recoveries reported that day. By the end of the period, the data shows about 49 million confirmed cases and over 1.2 million deaths worldwide. Because the dataset is large, covers many countries over time and contains real-world issues such as text-formatted dates, uneven coverage and negative correction values, it is well suited to a data manipulation project on cleaning, filtering, grouping and transforming data.

## <center> <h1 style="color:white; background-color:black; padding:8px;">About The Dataset</h1>
### The dataset used in this project, *COVID_Data_Basic.csv*, is a country-level COVID-19 time series with 54,522 records for 194 countries and regions, with one row for each country on each date. It covers 31 December 2019 to 6 November 2020, which includes the first wave of the pandemic and the start of the second. It has nine columns: a row index, the country name, the date, three cumulative totals (confirmed cases, deaths and recoveries) and three daily figures (new cases, new deaths and new recoveries). By the end of the period, the data shows roughly 49 million confirmed cases and 1.24 million deaths worldwide. There are no missing values, but the data needs cleaning because the dates are stored as text, some country names are inconsistent, coverage is uneven across countries and a few daily values are negative. This makes it a realistic dataset for practising data manipulation tasks such as cleaning, filtering, grouping and creating calculated columns.

## <center> <h1 style="color:white; background-color:black; padding:8px;">Abstract</h1>
### This project applies data manipulation techniques to a country-level COVID-19 dataset (*COVID_Data_Basic.csv*) containing 54,522 records for 194 countries and regions from 31 December 2019 to 6 November 2020. Each record holds the date, the cumulative confirmed cases, deaths and recoveries, and the new cases, deaths and recoveries reported that day. The data has no missing values, but it needs cleaning because the dates are stored as text, some country names are inconsistent, coverage is uneven across countries and a few daily values are negative. The project cleans and transforms the data, then uses filtering, grouping and aggregation to summarise how the pandemic developed across countries and over time. By the end of the period, the dataset shows roughly 49 million confirmed cases and 1.24 million deaths worldwide.

## <center> <h2 style="color:white; background-color:black; padding:8px;">Column Description</h2>

 <table style="border-collapse:collapse; width:100%; text-align:left;">
  <thead>
    <tr style="background-color:black; color:white;">
      <th style="padding:8px; border:1px solid #888;">Column</th>
      <th style="padding:8px; border:1px solid #888;">Data Type</th>
      <th style="padding:8px; border:1px solid #888;">Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="padding:8px; border:1px solid #888;"><b>Country</b></td>
      <td style="padding:8px; border:1px solid #888;">Text</td>
      <td style="padding:8px; border:1px solid #888;">Name of the country or region</td>
    </tr>
    <tr>
      <td style="padding:8px; border:1px solid #888;"><b>Date</b></td>
      <td style="padding:8px; border:1px solid #888;">Text (YYYY-MM-DD)</td>
      <td style="padding:8px; border:1px solid #888;">Date of the record</td>
    </tr>
    <tr>
      <td style="padding:8px; border:1px solid #888;"><b>Confirmed</b></td>
      <td style="padding:8px; border:1px solid #888;">Integer</td>
      <td style="padding:8px; border:1px solid #888;">Total confirmed cases up to that date</td>
    </tr>
    <tr>
      <td style="padding:8px; border:1px solid #888;"><b>Death</b></td>
      <td style="padding:8px; border:1px solid #888;">Integer</td>
      <td style="padding:8px; border:1px solid #888;">Total deaths up to that date</td>
    </tr>
    <tr>
      <td style="padding:8px; border:1px solid #888;"><b>Recovered</b></td>
      <td style="padding:8px; border:1px solid #888;">Integer</td>
      <td style="padding:8px; border:1px solid #888;">Total recoveries up to that date</td>
    </tr>
    <tr>
      <td style="padding:8px; border:1px solid #888;"><b>newConfirmed</b></td>
      <td style="padding:8px; border:1px solid #888;">Integer</td>
      <td style="padding:8px; border:1px solid #888;">New confirmed cases reported on that day</td>
    </tr>
    <tr>
      <td style="padding:8px; border:1px solid #888;"><b>newDeath</b></td>
      <td style="padding:8px; border:1px solid #888;">Integer</td>
      <td style="padding:8px; border:1px solid #888;">New deaths reported on that day</td>
    </tr>
    <tr>
      <td style="padding:8px; border:1px solid #888;"><b>newRecovered</b></td>
      <td style="padding:8px; border:1px solid #888;">Integer</td>
      <td style="padding:8px; border:1px solid #888;">New recoveries reported on that day</td>
    </tr>
  </tbody>
</table>

In [2]:
import pandas as pd

In [3]:
import warnings
warnings.filterwarnings("ignore")

In [4]:
df = pd.read_csv("COVID-19.csv")

In [5]:
df.head()

,Unnamed: 0,Country,Date,Confirmed,Death,Recovered,newConfirmed,newDeath,newRecovered
0,1,Afghanistan,2019-12-31,0,0,0,0,0,0
1,2,Afghanistan,2020-01-01,0,0,0,0,0,0
2,3,Afghanistan,2020-01-02,0,0,0,0,0,0
3,4,Afghanistan,2020-01-03,0,0,0,0,0,0
4,5,Afghanistan,2020-01-04,0,0,0,0,0,0


In [5]:
df.tail()

,Unnamed: 0,Country,Date,Confirmed,Death,Recovered,newConfirmed,newDeath,newRecovered
54517,55096,Zimbabwe,2020-11-02,8389,245,7939,15,2,12
54518,55097,Zimbabwe,2020-11-03,8410,246,7942,21,1,3
54519,55098,Zimbabwe,2020-11-04,8427,248,7967,17,2,25
54520,55099,Zimbabwe,2020-11-05,8444,248,7975,17,0,8
54521,55100,Zimbabwe,2020-11-06,8471,250,7983,27,2,8


In [6]:
df.drop(columns="Unnamed: 0",axis=1,inplace=True)

In [7]:
df

,Country,Date,Confirmed,Death,Recovered,newConfirmed,newDeath,newRecovered
0,Afghanistan,2019-12-31,0,0,0,0,0,0
1,Afghanistan,2020-01-01,0,0,0,0,0,0
2,Afghanistan,2020-01-02,0,0,0,0,0,0
3,Afghanistan,2020-01-03,0,0,0,0,0,0
4,Afghanistan,2020-01-04,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...
54517,Zimbabwe,2020-11-02,8389,245,7939,15,2,12
54518,Zimbabwe,2020-11-03,8410,246,7942,21,1,3
54519,Zimbabwe,2020-11-04,8427,248,7967,17,2,25
54520,Zimbabwe,2020-11-05,8444,248,7975,17,0,8


In [8]:
df.shape

(54522, 8)

In [9]:
df.size

436176

In [10]:
df.columns

Index(['Country', 'Date', 'Confirmed', 'Death', 'Recovered', 'newConfirmed',
       'newDeath', 'newRecovered'],
      dtype='object')

In [11]:
df.dtypes

Country         object
Date            object
Confirmed        int64
Death            int64
Recovered        int64
newConfirmed     int64
newDeath         int64
newRecovered     int64
dtype: object

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54522 entries, 0 to 54521
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Country       54522 non-null  object
 1   Date          54522 non-null  object
 2   Confirmed     54522 non-null  int64 
 3   Death         54522 non-null  int64 
 4   Recovered     54522 non-null  int64 
 5   newConfirmed  54522 non-null  int64 
 6   newDeath      54522 non-null  int64 
 7   newRecovered  54522 non-null  int64 
dtypes: int64(6), object(2)
memory usage: 3.3+ MB


In [13]:
df.describe()

,Confirmed,Death,Recovered,newConfirmed,newDeath,newRecovered
count,5.452200e+04,54522.000000,5.452200e+04,54522.000000,54522.000000,54522.000000
mean,7.293357e+04,2559.091981,4.584383e+04,904.639705,22.795734,595.733997
std,4.540187e+05,13334.101849,3.002110e+05,5118.383587,122.875597,4336.617112
min,0.000000e+00,0.000000,0.000000e+00,-10034.000000,-1918.000000,-16298.000000
25%,2.400000e+01,0.000000,5.000000e+00,0.000000,0.000000,0.000000
50%,1.368500e+03,26.000000,6.010000e+02,11.000000,0.000000,1.000000
75%,1.582800e+04,319.000000,7.930500e+03,193.000000,3.000000,82.000000
max,9.733816e+06,236073.000000,7.819886e+06,126480.000000,4143.000000,388340.000000


In [14]:
df.isnull().sum()

Country         0
Date            0
Confirmed       0
Death           0
Recovered       0
newConfirmed    0
newDeath        0
newRecovered    0
dtype: int64

In [15]:
df.isnull().sum().sum()

np.int64(0)

Since there is no null values.so we don't need to handle the null values. .

In [16]:
df.duplicated().sum()

np.int64(0)

There is no unique records.

In [17]:
df["Date"].max()

'2020-11-06'

## Observation :
#### The dataset ends on **6 November 2020**, so all totals and rankings are as of that date.

In [18]:
df["Date"].min()

'2019-12-31'

## Observation :
#### The dataset starts on **31 December 2019**, when the first case was reported.

In [21]:
df.groupby("Country")["Death"].max().sort_values(ascending = False)

Country
US                       236073
Brazil                   162015
India                    125562
Mexico                    94323
United Kingdom            48565
                          ...  
Saint Kitts and Nevis         0
Bhutan                        0
Marshall Islands              0
Cambodia                      0
Dominica                      0
Name: Death, Length: 194, dtype: int64

## Observation :
#### The country with the most death is in **US**.

In [30]:
df.groupby("Country")["Confirmed"].max().sort_values(ascending = False)

Country
US                  9733816
India               8462080
Brazil              5631181
Russia              1720063
France              1709716
                     ...   
Western Sahara           10
MS Zaandam                9
Bahamas, The              3
Marshall Islands          2
Gambia, The               1
Name: Confirmed, Length: 194, dtype: int64

## Observation :
#### The country with most confirmed cases is in **US**.

In [37]:
df.groupby("Country")["Confirmed"].max().sort_values(ascending = False).head(5)

Country
US        9733816
India     8462080
Brazil    5631181
Russia    1720063
France    1709716
Name: Confirmed, dtype: int64

## Observation :
#### The top 5 most confirmed cases.

In [35]:
df["Confirmed"].max()

9733816

## Obsevation :
#### The maximum number of confirmed cases is **9733816**.

In [8]:
df.groupby("Country")["Recovered"].max().sum()

np.int64(32496531)

## Observation :
#### The Total Recovered in all countries is **32496531**.

In [53]:
df.groupby("Country")["Recovered"].max().sort_values(ascending = False)

Country
India               7819886
Brazil              5118156
US                  3810791
Russia              1288096
Argentina           1042237
                     ...   
Bahamas, The              0
Gambia, The               0
Martinique                0
Marshall Islands          0
MS Zaandam                0
Name: Recovered, Length: 194, dtype: int64

## Observation :
#### The Country with the most Recovered is in **India**.

In [56]:
df["newConfirmed"].sum()

np.int64(49322766)

## Observation :
#### The Total newConfirmed cases in all over the Country is **49322766**.

In [63]:
df.groupby("Country")["newConfirmed"].max().sort_values(ascending = False)

Country
US                  126480
France              124905
India                97894
Brazil               69074
Spain                55019
                     ...  
Laos                     3
Bahamas, The             2
Marshall Islands         2
Gambia, The              1
Diamond Princess         0
Name: newConfirmed, Length: 194, dtype: int64

## Observation :
#### The country with most newConfirmed cases is in **US**.

In [8]:
df.groupby("Country")["newDeath"].max().sort_values(ascending = False)

Country
Peru               4143
Ecuador            3852
Argentina          3351
Mexico             2789
US                 2609
                   ... 
Mongolia              0
Solomon Islands       0
Dominica              0
Bahamas, The          0
Eritrea               0
Name: newDeath, Length: 194, dtype: int64

## Observation :
#### The Country with most newDeath was in **Peru** with **4143**.

In [6]:
df.groupby("Country")["newRecovered"].max().sort_values(ascending = False)

Country
Brazil              388340
US                  103921
India               101468
Argentina            73147
Romania              44991
                     ...  
MS Zaandam               0
Marshall Islands         0
Martinique               0
Bahamas, The             0
Gambia, The              0
Name: newRecovered, Length: 194, dtype: int64

## Observation :
#### The Country with the most newRecovered is **Brazil**.

In [9]:
df.to_csv("New_COVID-19.csv" ,index = False)

## <center> <h1 style="color:white; background-color:black; padding:8px;">Conclusion</h1>
### This project cleaned and analysed a country-level COVID-19 dataset of 54,522 records covering 194 countries and regions, from 31 December 2019 to 6 November 2020. After dropping the unnecessary index column, the data had no missing values and no duplicate records. The analysis shows that cases were concentrated in a few countries. The US had the most confirmed cases (9,733,816) and the most deaths (236,073), followed by India (8,462,080 cases) and Brazil (5,631,181 cases), with Russia and France next. India had the most recoveries (7,819,886), followed by Brazil. The highest single-day increase in cases was in the US (126,480), and Peru had the highest single-day death count (4,143), which may reflect a one-off revision of earlier figures. Across all countries, the daily new cases add up to about 49.3 million confirmed cases by 6 November 2020. Overall, the pandemic hit a small number of large countries hardest, and the data was still showing rising cases at the end of the period.